evm_arithmetization/src/curve_pairings.rs

```diff
@@ -150,366 +141,937 @@ where
-            y4 = y4 * sq;
+/// Generates a sparse, random Fp12 element.
+pub(crate) fn gen_fp12_sparse<F, R: Rng + ?Sized>(rng: &mut R) -> Fp12<F>
+where
+    F: FieldExt,
+    Fp2<F>: Adj,
+    Standard: Distribution<F>,
+{
+    sparse_embed::<F>(rng.gen::<F>(), rng.gen::<Fp2<F>>(), rng.gen::<Fp2<F>>())
+}
+
+pub mod bn254 {
+    use super::*;
+
+    /// The BN curve consists of pairs
+    ///     (x, y): (BN254, BN254) | y^2 = x^3 + 3
+    // with generator given by (1, 2)
+    impl CyclicGroup for CurveAff<BN254> {
+        const GENERATOR: CurveAff<BN254> = CurveAff {
+            x: BN254 { val: U256::one() },
+            y: BN254 {
+                val: U256([2, 0, 0, 0]),
+            },
+        };
+    }
+
+    /// The twisted curve consists of pairs
+    ///     (x, y): (Fp2<BN254>, Fp2<BN254>) | y^2 = x^3 + 3/(9 + i)
+    /// with generator given as follows
+    impl CyclicGroup for CurveAff<Fp2<BN254>> {
+        const GENERATOR: CurveAff<Fp2<BN254>> = CurveAff {
+            x: Fp2 {
+                re: BN254 {
+                    val: U256([
+                        0x46debd5cd992f6ed,
+                        0x674322d4f75edadd,
+                        0x426a00665e5c4479,
+                        0x1800deef121f1e76,
+                    ]),
+                },
+                im: BN254 {
+                    val: U256([
+                        0x97e485b7aef312c2,
+                        0xf1aa493335a9e712,
+                        0x7260bfb731fb5d25,
+                        0x198e9393920d483a,
+                    ]),
+                },
+            },
+            y: Fp2 {
+                re: BN254 {
+                    val: U256([
+                        0x4ce6cc0166fa7daa,
+                        0xe3d1e7690c43d37b,
+                        0x4aab71808dcb408f,
+                        0x12c85ea5db8c6deb,
+                    ]),
+                },
+                im: BN254 {
+                    val: U256([
+                        0x55acdadcd122975b,
+                        0xbc4b313370b38ef3,
+                        0xec9e99ad690c3395,
+                        0x090689d0585ff075,
+                    ]),
+                },
+            },
+        };
+    }
+
+    /// The sloped line function for doubling a point.
+    pub(crate) fn tangent(p: CurveAff<BN254>, q: CurveAff<Fp2<BN254>>) -> Fp12<BN254> {
+        let cx = -BN254::new(3) * p.x * p.x;
+        let cy = BN254::new(2) * p.y;
+        sparse_embed::<BN254>(p.y * p.y - BN254::new(9), q.x * cx, q.y * cy)
+    }
+
+    /// The sloped line function for adding two points.
+    pub(crate) fn chord(
+        p1: CurveAff<BN254>,
+        p2: CurveAff<BN254>,
+        q: CurveAff<Fp2<BN254>>,
+    ) -> Fp12<BN254> {
+        let cx = p2.y - p1.y;
+        let cy = p1.x - p2.x;
+        sparse_embed::<BN254>(p1.y * p2.x - p2.y * p1.x, q.x * cx, q.y * cy)
+    }
+
+    // The tate pairing takes a point each from the curve and its twist and outputs
```

Comment:
```suggestion
    // The tate pairing takes a point from the curve and its twist and outputs
```
Not sure, but otherwise sounds odd.